<!-- 
The MIT License (MIT)
Copyright (c) 2016 GRVC University of Seville

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 
-->
<launch>

    <arg name="multi" default="false"/>
    <arg name="ns_prefix" default="uav_"/>
    <arg name="mode" default="sitl"/>
    <arg name="pose_frame_id" default=""/><!-- To publish ual/pose in other frame than local, e.g. map -->
    <arg name="ual_server" default="on"/>

    <!-- Load frame transforms for map and UAVs -->
    <rosparam file="$(find px4_bringup)/config/simulation_params.yaml" command="load" />

    <group if="$(eval mode=='sitl')">
        <!-- Launch Gazebo simulation TODO: world location? -->
        <node pkg="px4_bringup" type="launch_gzworld.py" name="gazebo_world" output="screen"
        args="-physics=ode -world=$(find px4_bringup)/config/empty_light.world"/>
    </group>

    <!-- Launch robots. Argument mode can be:
        - sitl: for software in the loop simulation
        - serial: for serial connection to px4
        - udp: for udp connection to px4
    -->
    <!-- Launch robot id=1 -->
    <include file="$(find px4_bringup)/launch/spawn_robot.launch">
        <arg name="id" value="1"/>
        <arg name="material" value="Orange"/>
        <arg name="mode" value="$(arg mode)"/>
        <arg name="ns_prefix" value="$(arg ns_prefix)"/>
    </include>

    <group if="$(arg multi)">
        <!-- Launch robot id=2 -->
        <include file="$(find px4_bringup)/launch/spawn_robot.launch">
            <arg name="id" value="2"/>
            <arg name="material" value="Indigo"/>
            <arg name="mode" value="sitl"/>
            <arg name="ns_prefix" value="$(arg ns_prefix)"/>
        </include>
        <!-- Launch robot id=3 -->
        <include file="$(find px4_bringup)/launch/spawn_robot.launch">
            <arg name="id" value="3"/>
            <arg name="material" value="ZincYellow"/>
            <arg name="mode" value="sitl"/>
            <arg name="ns_prefix" value="$(arg ns_prefix)"/>
        </include>
    </group>

    <!-- Launch UAL server for uav id=1 -->
    <node pkg="uav_abstraction_layer" type="server" name="server_1" output="screen">
        <param name="uav_id" value="1"/>
        <param name="ns_prefix" value="$(arg ns_prefix)"/>
        <param name="pose_frame_id" value="$(arg pose_frame_id)"/>
        <param name="ual_server" value="$(arg ual_server)"/>
    </node>

    <group if="$(arg multi)">
        <!-- Launch UAL server for uav id=2 -->
        <node pkg="uav_abstraction_layer" type="server" name="server_2" output="screen">
            <param name="uav_id" value="2"/>
            <param name="ns_prefix" value="$(arg ns_prefix)"/>
            <param name="pose_frame_id" value="$(arg pose_frame_id)"/>
            <param name="ual_server" value="$(arg ual_server)"/>
        </node>
        <!-- Launch UAL server for uav id=3 -->
        <node pkg="uav_abstraction_layer" type="server" name="server_3" output="screen">
            <param name="uav_id" value="3"/>
            <param name="ns_prefix" value="$(arg ns_prefix)"/>
            <param name="pose_frame_id" value="$(arg pose_frame_id)"/>
            <param name="ual_server" value="$(arg ual_server)"/>
        </node>
    </group>

</launch>
